File: framework/fit/python/fit_flowable/operation/abstract_operation.py
# -- encoding: utf-8 --
# Copyright (c) Huawei Technologies Co., Ltd. 2024-2024. All rights reserved.
"""
功 能：同时是订阅者和订阅关系的抽象父类。
"""
from abc import abstractmethod
from typing import Optional

from fit_flowable.core.exception import FlowableException
from fit_flowable.core.subscriber import Subscriber
from fit_flowable.core.subscription import Subscription
from fit_flowable.util.concurrent import AtomicBool


class AbstractOperation(Subscriber, Subscription):
    """
    同时是订阅者和订阅关系的抽象父类。
    """

    def __init__(self, next_subscriber: Subscriber):
        """
        通过下游订阅者构造。

        :param next_subscriber: 下游的订阅者。
        """
        if next_subscriber is None:
            raise ValueError("the next subscriber cannot be null.")
        self._next_subscriber: Subscriber = next_subscriber
        self._pre_subscription: Optional[Subscription] = None
        self._completed: AtomicBool = AtomicBool()
        self._failed: AtomicBool = AtomicBool()
        self._cancelled: AtomicBool = AtomicBool()

    @abstractmethod
    def do_consume(self, data) -> None:
        """
        用于定义消费订阅的数据时的实际行为。

        :param data: 待消费的数据。
        """
        pass

    def do_fail(self, cause: Exception) -> None:
        """
        用于定义处理异常终结信号时的实际行为。

        :param cause: 待处理的异常。
        """
        self._next_subscriber.fail(cause)

    def do_complete(self) -> None:
        """
        用于定义数据发送完成时的实际行为。
        """
        self._next_subscriber.complete()

    def get_next_subscriber(self) -> Subscriber:
        """
        获取下游的订阅者。

        :return: 获取到的下游订阅者。
        """
        return self._next_subscriber

    def get_pre_subscription(self) -> Subscription:
        """
        获取上游订阅关系。

        :return: 获取到的订阅关系。
        """
        if self._pre_subscription is None:
            raise FlowableException("could not access property subscription before on subscribed.")
        return self._pre_subscription

    def get_completed(self) -> bool:
        """
        获取是否已经正常终结。

        :return: 如果已经正常终结则返回 True，否则，返回 False。
        """
        return self._completed.get()

    def get_failed(self) -> bool:
        """
        获取是否已经异常终结。

        :return: 如果已经异常终结则返回 True，否则，返回 False。
        """
        return self._failed.get()

    def get_cancelled(self) -> bool:
        """
        获取是否已经取消。

        :return: 如果已经取消则返回 True，否则，返回 False。
        """
        return self._cancelled.get()

    def on_subscribed(self, subscription: Subscription) -> None:
        if self._pre_subscription is not None:
            raise FlowableException("on_subscribed could not called twice.")
        self._pre_subscription = subscription
        self._next_subscriber.on_subscribed(self)

    def consume(self, data) -> None:
        if self._is_terminated():
            return
        self.do_consume(data)

    def fail(self, cause: Exception) -> None:
        if self._is_terminated() or not self._failed.compare_and_set(False, True):
            return
        self.do_fail(cause)

    def complete(self) -> None:
        if self._is_terminated() or not self._completed.compare_and_set(False, True):
            return
        self.do_complete()

    def request(self, count: int) -> None:
        if self._is_terminated():
            return
        if count <= 0:
            raise ValueError(f"The number of elements to request must be positive. [count={count}]")
        self._pre_subscription.request(count)

    def cancel(self) -> None:
        if self._cancelled.compare_and_set(False, True):
            return
        self._pre_subscription.cancel()

    def _is_terminated(self) -> bool:
        return self._cancelled.get() or self._failed.get() or self._completed.get()

File: framework/fit/java/fit-reactor/src/main/java/modelengine/fitframework/flowable/operation/AbstractOperation.java
/*
 * Copyright (c) Huawei Technologies Co., Ltd. 2024-2024. All rights reserved.
 */

package modelengine.fitframework.flowable.operation;

import static modelengine.fitframework.inspection.Validation.greaterThan;
import static modelengine.fitframework.inspection.Validation.notNull;

import modelengine.fitframework.flowable.Subscriber;
import modelengine.fitframework.flowable.Subscription;
import modelengine.fitframework.flowable.subscriber.AbstractSubscriber;
import modelengine.fitframework.inspection.Nonnull;

import java.util.concurrent.atomic.AtomicBoolean;

/**
 * 表示同时是订阅者和订阅关系的抽象父类。
 *
 * @param <T> 表示当前类作为订阅者时，所订阅的数据类型的 {@link T}。
 * @param <R> 表示当前类作为订阅关系时，下游订阅者所订阅的数据类型的 {@link R}。
 * @author 季聿阶
 * @since 2024-02-09
 */
public abstract class AbstractOperation<T, R> extends AbstractSubscriber<T> implements Subscription {
    private final Subscriber<R> nextSubscriber;
    private final AtomicBoolean cancelled = new AtomicBoolean();

    /**
     * 使用下游的订阅者来初始化 {@link AbstractOperation}。
     *
     * @param nextSubscriber 表示下游的订阅者的 {@link Subscriber}{@code <}{@link R}{@code >}。
     */
    protected AbstractOperation(Subscriber<R> nextSubscriber) {
        this.nextSubscriber = notNull(nextSubscriber, "The next subscriber cannot be null.");
    }

    /**
     * 获取下游的订阅者。
     *
     * @return 表示下游的订阅者的 {@link Subscriber}{@code <}{@link R}{@code >}。
     */
    @Nonnull
    protected Subscriber<R> getNextSubscriber() {
        return this.nextSubscriber;
    }

    @Override
    protected void onSubscribed0(@Nonnull Subscription subscription) {
        this.nextSubscriber.onSubscribed(this);
    }

    @Override
    public void request(long count) {
        greaterThan(count, 0, "The number of elements to request must be positive. [count={0}]", count);
        if (this.cancelled.get()) {
            return;
        }
        this.request0(count);
    }

    /**
     * 请求指定数量的数据。
     * <p>当前请求没有被取消。</p>
     *
     * @param count 表示请求的数据的数量的 {@code long}。
     */
    protected void request0(long count) {
        this.getPreSubscription().request(count);
    }

    @Override
    public void cancel() {
        if (!this.cancelled.compareAndSet(false, true)) {
            return;
        }
        this.cancel0();
    }

    /**
     * 取消当前的订阅关系。
     * <p>默认调用上游订阅关系进行取消。</p>
     */
    protected void cancel0() {
        this.getPreSubscription().cancel();
    }

    @Override
    public boolean isCancelled() {
        return this.cancelled.get();
    }

    @Override
    protected void consume(@Nonnull Subscription subscription, T data) {
        if (this.isTerminated()) {
            return;
        }
        this.consume0(subscription, data);
    }

    /**
     * 消费订阅的数据。
     * <p>当前待消费的数据处于可消费的状态。</p>
     *
     * @param subscription 表示上游的订阅关系的 {@link Subscription}。
     * @param data 表示待消费的数据的 {@link T}。
     */
    protected abstract void consume0(Subscription subscription, T data);

    @Override
    protected void complete(@Nonnull Subscription subscription) {
        if (this.cancelled.get()) {
            return;
        }
        this.complete0(subscription);
    }

    /**
     * 消费正常终结信号。
     * <p>当前待消费的正常终结信号处于可消费的状态。</p>
     *
     * @param subscription 表示上游的订阅关系的 {@link Subscription}。
     */
    protected void complete0(Subscription subscription) {
        this.getNextSubscriber().complete();
    }

    @Override
    protected void fail(@Nonnull Subscription subscription, Exception cause) {
        if (this.cancelled.get()) {
            return;
        }
        this.fail0(subscription, cause);
    }

    /**
     * 消费异常终结信号。
     * <p>当前待消费的异常终结信号处于可消费的状态。</p>
     *
     * @param subscription 表示上游的订阅关系的 {@link Subscription}。
     * @param cause 表示待消费的异常 {@link Exception}
     */
    protected void fail0(Subscription subscription, Exception cause) {
        this.getNextSubscriber().fail(cause);
    }

    private boolean isTerminated() {
        return this.cancelled.get() || this.isCompleted() || this.isFailed();
    }
}
